File: LeetCode/Difficulty/Medium/py/98.validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def solve(self, root: TreeNode, minVal: int, maxVal: int) -> bool:
        if not root:
            return True
        return minVal <= root.val and root.val <= maxVal and self.solve(root.left, minVal, root.val-1) and self.solve(root.right, root.val+1, maxVal)

    
    def isValidBST(self, root: TreeNode) -> bool:
        return self.solve(root, -(1 << 31), (1 << 31) - 1)

File: LeetCode/Tag/Binary-Search-Tree/js/98.validate-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root, lowerBound = -Infinity, upperBound = Infinity) {
    if(!root) return true;

    if(lowerBound < root.val && root.val < upperBound){
        return isValidBST(root.left,lowerBound, root.val) && isValidBST(root.right,root.val, upperBound);
    } 
    else{
        return false;
    }
};
